packages/flutter_tools/lib/src/build_system/targets/font_subset.dart

```diff
@@ -0,0 +1,359 @@
+        continue;
+      }
+      final List<Map<String, dynamic>> fonts = _getList(
+        map['fonts'],
+        'FontManifest.json invalid: expected "fonts" to be a list of objects.',
+      );
+      if (fonts.length != 1) {
+        throw FontSubsetException._(
+          'This tool cannot process icon fonts with multiple fonts in a '
+          'single family.');
+      }
+      if (fonts.first['asset'] is! String) {
+        throw FontSubsetException._(
+          'FontManifest.json invalid: expected "asset" value to be a string, '
+          'got: ${map['assets']}.');
+      }
+      result[familyKey] = fonts.first['asset'] as String;
+    }
+    return result;
+  }
+
+  Future<Map<String, List<int>>> _findConstants(
+    File dart,
+    File constFinder,
+    File appDill,
+  ) async {
+    final List<String> cmd = <String>[
+      dart.path,
+      constFinder.path,
+      '--kernel-file', appDill.path,
+      '--class-library-uri', 'package:flutter/src/widgets/icon_data.dart',
+      '--class-name', 'IconData',
+    ];
+    _logger.printTrace('Running command: ${cmd.join(' ')}');
+    final ProcessResult constFinderProcessResult = await _processManager.run(cmd);
+
+    if (constFinderProcessResult.exitCode != 0) {
+      throwToolExit('ConstFinder failure: ${constFinderProcessResult.stderr}');
+    }
+    final dynamic jsonDecode = json.decode(constFinderProcessResult.stdout as String);
+    if (jsonDecode is! Map<String, dynamic>) {
+      throw FontSubsetException._(
+        'Invalid ConstFinder output: expected a top level JavaScript object, '
+        'got $jsonDecode.');
+    }
+    final Map<String, dynamic> constFinderMap = jsonDecode as Map<String, dynamic>;
+    final _ConstFinderResult constFinderResult = _ConstFinderResult(constFinderMap);
+    if (constFinderResult.hasNonConstantLocations) {
+      _logger.printError('This application cannot tree shake icons fonts. '
+                         'It has non-constant instances of IconData at the '
+                         'following locations:', emphasis: true);
+      for (final Map<String, dynamic> location in constFinderResult.nonConstantLocations) {
+        _logger.printError(
+          '- ${location['file']}:${location['line']}:${location['column']}',
+          indent: 2,
+          hangingIndent: 4,
+        );
+      }
+      throwToolExit('Avoid non-constant invocations of IconData or try to '
+                    'build again with --no-tree-shake-icons.');
+    }
+    return _parseConstFinderResult(constFinderResult);
+  }
+
+  Map<String, List<int>> _parseConstFinderResult(_ConstFinderResult consts) {
+    final Map<String, List<int>> result = <String, List<int>>{};
+    for (final Map<String, dynamic> iconDataMap in consts.constantInstances) {
+      if ((iconDataMap['fontPackage'] ?? '') is! String || // Null is ok here.
+           iconDataMap['fontFamily'] is! String ||
+           iconDataMap['codePoint'] is! int) {
+        throwToolExit(
```

Comment:
Is this an internal error, or something the user can do something about?

Reply:
Changed to exception - this is not a user error.